src/main/java/org/broadinstitute/hellbender/tools/spark/validation/CompareDuplicates.java

```diff
@@ -0,0 +1,203 @@
+            }
+            return !Utils.isNonPrimary(v1);
+        });
+        secondReads.foreach(GATKRead::clearAttributes);
+
+
+        // Start by verifying that we have same number of reads and duplicates in each BAM.
+        long firstBamSize = firstReads.count();
+        JavaRDD<GATKRead> firstDupes = firstReads.filter(GATKRead::isDuplicate);
+        long firstDupesCount = firstDupes.count();
+
+        long secondBamSize = secondReads.count();
+        JavaRDD<GATKRead> secondDupes = secondReads.filter(GATKRead::isDuplicate);
+        long secondDupesCount = secondDupes.count();
+
+        if (firstBamSize != secondBamSize) {
+            throw new UserException("input bams have different numbers of mapped reads: "
+                    + firstBamSize + "," + secondBamSize);
+        }
+        System.out.println("processing bams with " + firstBamSize + " mapped reads");
+
+
+        System.out.println("first and second: " + firstDupesCount + "," + secondDupesCount);
+        if (firstDupesCount != secondDupesCount) {
+            System.out.println("BAMs have different number of total duplicates: " + firstDupesCount + "," + secondDupesCount);
+        }
+
+        // Group the reads of each BAM by MarkDuplicates key, then pair up the the reads for each BAM.
+        JavaPairRDD<String, GATKRead> firstKeyed = firstReads.mapToPair(read -> new Tuple2<>(Utils.groupKey(read), read));
+        JavaPairRDD<String, GATKRead> secondKeyed = secondReads.mapToPair(read -> new Tuple2<>(Utils.groupKey(read), read));
+        JavaPairRDD<String, Tuple2<Iterable<GATKRead>, Iterable<GATKRead>>> cogroup = firstKeyed.cogroup(secondKeyed);
+
+        Broadcast<SAMFileHeader> broadcastHeader = ctx.broadcast(getHeaderForReads());
+
+        // Produces an RDD of MatchTypes, e.g., EQUAL, DIFFERENT_REPRESENTATIVE_READ, etc. per MarkDuplicates key,
+        // which is approximately start position x strand.
+        JavaRDD<Utils.MatchType> tagged = cogroup.map(v1 -> {
+            SAMFileHeader header = broadcastHeader.getValue();
+
+            Iterable<GATKRead> iFirstReads = v1._2()._1();
+            Iterable<GATKRead> iSecondReads = v1._2()._2();
+
+            return Utils.getDupes(iFirstReads, iSecondReads, header);
+        });
+
+        Map<Utils.MatchType, Integer> tagCountMap = tagged.mapToPair(v1 ->
+                new Tuple2<>(v1, 1)).reduceByKey((v1, v2) -> v1 + v2).collectAsMap();
+
+        if (tagCountMap.get(Utils.MatchType.SIZE_UNEQUAL) != null) {
+            throw new UserException("The number of reads by the MarkDuplicates key were unequal, indicating that the BAMs are not the same");
+        }
+        if (tagCountMap.get(Utils.MatchType.READ_MISMATCH) != null) {
+            throw new UserException("The reads grouped by the MarkDuplicates key were not the same, indicating that the BAMs are not the same");
```

Comment:
this would be so much more informative if it printed an example problematic read (or a group of them) - right now i'm still left with the work of actually finding the difference